File: 주영/Array/118. Pascal's Triangle.js
var generate = function (numRows) {
  let answer = [];
  let preLine = [];

  answer.push([1]);

  while (answer.length < numRows) {
    let currentLine = [];
    currentLine[0] = 1; // 처음에 push 1 해주고

    for (let i = 0; i < preLine.length; i++) {
      currentLine[i + 1] = preLine[i] + preLine[i + 1];
    }
    currentLine.push(1); // 마지막 push 1 해주기
    answer.push(currentLine);
    // console.log('answer: ', answer);
    preLine = currentLine; // 다음 행으로 넘어가기
  }
  return answer;
};

let numRows = 5;
generate(numRows);

File: P/118_Pascal's_Triangle.java
//Time: O(n^2) Space: O(n^2) where n = numRows
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> list = new ArrayList<List<Integer>>();
        
        for(int i = 0; i < numRows; i++) {
            List<Integer> arr = new ArrayList<>();
            for(int j = 0; j <= i; j++) {
                if(j == 0 || j == i) {
                    arr.add(1);
                } else {
                    int num = list.get(i - 1).get(j - 1) + list.get(i - 1).get(j);
                    arr.add(num);
                }
            }
            list.add(arr);
        }
        return list;
    }
}